category/traversable/instances.lean

```diff
@@ -0,0 +1,249 @@
+/-
+Copyright (c) 2018 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+
+Instances of `traversable` for types from the core library
+-/
+
+import category.traversable.basic
+import category.basic
+import category.functor
+import category.applicative
+
+universe variables u v
+
+open function
+
+section identity
+
+open function functor
+
+variables {f f' : Type u → Type u}
+variables [applicative f] [applicative f']
+
+def id.traverse {α β : Type u} (g : α → f β) : id α → f (id β) := g
+
+instance : traversable id :=
+⟨ @id.traverse ⟩
+
+variables [is_lawful_applicative f] [is_lawful_applicative f']
+variables {α β γ : Type u}
+
+lemma id.id_traverse (x : id α) :
+  id.traverse id.mk x = x :=
+by refl
+lemma id.comp_traverse (g : α → f β) (h : β → f' γ) (x : id α) :
+  id.traverse (comp.mk ∘ map h ∘ g) x =
+  comp.mk (id.traverse h <$> id.traverse g x) :=
+by simp! [id.traverse,functor.map_id] with functor_norm
+
+lemma id.map_traverse
+   (g : α → f' β) (f : β → γ)
+   (x : id α) :
+  map f <$> id.traverse g x = id.traverse (map f ∘ g) x :=
+by simp [map,id_bind,id.traverse] with functor_norm
+
+lemma id.traverse_map
+   (g : α → β) (f : β → f' γ)
+   (x : id α) :
+  id.traverse f (g <$> x) = id.traverse (f ∘ g) x :=
+by simp [map,id_bind,id.traverse] with functor_norm
+
+variable (eta : applicative_transformation f f')
+
+lemma id.naturality {α β : Type u}
+  (F : α → f β) (x : id α) :
+  eta (id.traverse F x) = id.traverse (@eta _ ∘ F) x :=
+by simp! [id.traverse] with functor_norm; refl
+
+end identity
+
+instance : is_lawful_traversable id :=
+{ id_traverse := λ α x, @id.id_traverse α x,
+  comp_traverse := @id.comp_traverse,
+  map_traverse := @id.map_traverse,
+  traverse_map := @id.traverse_map,
+  naturality := @id.naturality }
+
+section option
+
+open function functor
+
+variables {f f' : Type u → Type u}
+variables [applicative f] [applicative f']
+
+protected def option.traverse {α β : Type u} (g : α → f β) : option α → f (option β)
```

Comment:
This function should be defined on different universes. I know that the traversable instance can't use different universes, but the function can still be used directly. (This is often the situation with `option.map`.)

Reply:
Good point. I also tried to make the lemmas more polymorphic but I sense it can't be done because it relies on arbitrary `applicative`s.